lib/typing/mongo/bson.go

```diff
@@ -32,133 +27,101 @@ func JSONEToMap(val []byte) (map[string]any, error) {
-	if !isOk {
-		return bsoncodec.ValueEncoderError{Name: "ObjectIDEncodeValue not Binary", Types: []reflect.Type{tBinary}, Received: val}
-	}
-
-	switch s.Subtype {
+func bsonBinaryValueToMap(value primitive.Binary) (any, error) {
+	switch value.Subtype {
 	case
 		bson.TypeBinaryUUIDOld,
 		bson.TypeBinaryUUID:
-		parsedUUID, err := uuid.FromBytes(s.Data)
+		parsedUUID, err := uuid.FromBytes(value.Data)
 		if err != nil {
-			return err
+			return nil, err
 		}
-		return vw.WriteString(parsedUUID.String())
+
+		return parsedUUID.String(), nil
 	default:
-		return vw.WriteBinaryWithSubtype(s.Data, s.Subtype)
+		return map[string]any{
+			"$binary": map[string]any{
+				"base64":  base64.StdEncoding.EncodeToString(value.Data),
+				"subType": fmt.Sprintf("%02x", value.Subtype),
+			},
+		}, nil
 	}
 }
 
-func timestampEncodeValue(_ bsoncodec.EncodeContext, vw bsonrw.ValueWriter, val reflect.Value) error {
-	if !val.IsValid() || val.Type() != tTimestamp {
-		return bsoncodec.ValueEncoderError{Name: "ObjectIDEncodeValue", Types: []reflect.Type{tBinary}, Received: val}
-	}
-
-	s, isOk := val.Interface().(primitive.Timestamp)
-	if !isOk {
-		return bsoncodec.ValueEncoderError{Name: "ObjectIDEncodeValue not Binary", Types: []reflect.Type{tBinary}, Received: val}
-	}
+func bsonValueToGoValue(value any) (any, error) {
+	switch v := value.(type) {
+	case primitive.DateTime:
+		return time.Unix(0, int64(v)*int64(time.Millisecond)).UTC().Format(ext.ISO8601), nil
+	case primitive.ObjectID:
+		return v.Hex(), nil
+	case primitive.Binary:
+		return bsonBinaryValueToMap(v)
+	case primitive.Decimal128:
+		potentialFloat, err := strconv.ParseFloat(v.String(), 64)
+		if err != nil {
+			return nil, err
+		}
 
-	return vw.WriteString(time.Unix(int64(s.T), 0).UTC().Format(ext.ISO8601))
-}
+		if v.String() == fmt.Sprint(potentialFloat) {
+			return potentialFloat, nil
+		}
 
-func createCustomRegistry() *bsoncodec.Registry {
-	rb := bson.NewRegistry()
-	rb.RegisterTypeEncoder(tDateTime, bsoncodec.ValueEncoderFunc(dateTimeEncodeValue))
-	rb.RegisterTypeEncoder(tOID, bsoncodec.ValueEncoderFunc(objectIDEncodeValue))
-	rb.RegisterTypeEncoder(tBinary, bsoncodec.ValueEncoderFunc(binaryEncodeValue))
-	rb.RegisterTypeEncoder(tDecimal, bsoncodec.ValueEncoderFunc(decimalEncodeValue))
-	rb.RegisterTypeEncoder(tTimestamp, bsoncodec.ValueEncoderFunc(timestampEncodeValue))
-	return rb
+		// If the string representation of the Decimal128 is not equal to the float representation
+		// then we return the string representation
+		return v.String(), nil
+	case primitive.Timestamp:
+		return time.Unix(int64(v.T), 0).UTC().Format(ext.ISO8601), nil
+	case bson.D:
+		return bsonDocToMap(v)
+	case bson.A:
+		return bsonArrayToSlice(v)
+	case primitive.MaxKey:
+		return map[string]any{"$maxKey": 1}, nil
+	case primitive.MinKey:
+		return map[string]any{"$minKey": 1}, nil
+	case primitive.JavaScript:
+		return map[string]any{"$code": fmt.Sprint(v)}, nil
```

Comment:
Can probably do `string(v)` since `primitive.JavaScript` is just a `type JavaScript string`.